### DISCUSSION TITLE
Is it possible to determine whether a SyntaxNode or Token originates from the source text or is expanded from macros?

### DISCUSSION THREAD

**Discussion by roife** [MikePopoloski/slang#1172] [Q&A] [answered]
Hi!👋 Thank you for your dedication to maintaining this project!

I am currently learning the API of slang and have written some tests, and I noticed that slang automatically expands macros when generating a SyntaxTree.

Is it possible to obtain a SyntaxTree that has not been preprocessed, or is there a method to distinguish whether a SyntaxNode originates from text or it is expanded from a macro?

**Comment by MikePopoloski** [comment, marked as answer]
You cannot obtain a non-preprocessed SyntaxTree; it's not possible to parse SystemVerilog without first expanding macros.

You *can* check whether any particular token comes from a macro or not by looking at its SourceLocation and using SourceManager::isMacroLoc to check.
